Spock testing

Kleamens opened the pull request:
Added dependencies for spock testing 

Added unit tests for the data classes 

Added integration tests for the service layers

Kleamens committed:
feat: spock dependencies + base structure

Since in most cases you don't need several unit testing frameworks, the part with the JUnit tests is commented

Kleamens committed:
feat: added unit tests + service tests with cleanup and setup

Addeed unit tests + data cleanup on and setup before every test with a global cleanup and setup

jankoscak commented on moviesAPI/build.gradle:
```diff
@@ -34,14 +35,19 @@ dependencies {
+//	testImplementation 'org.junit.jupiter:junit-jupiter-engine'
 
-	testImplementation 'org.hamcrest:hamcrest:2.2'
+//	testImplementation 'org.hamcrest:hamcrest:2.2'
```
🐛 Nezakomentovavat, ale mazat. Je to pak neprehladne. O pul roku budete resit, proc je to zakomentovane a k cemu to vlastne je.

jankoscak commented on moviesAPI/src/test/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/MoviesApiApplicationTests.java:
```diff
@@ -1,13 +1,13 @@
+//	void contextLoads() {
+//	}
+//
+//}
```
🐛 Opet zmazat a nebo nechat tak. Spock i Junit testy muzou bezet vedle sebe a tudis dokud nebudete mit komplet spock testy, tak muzete mit i junit testy a zmazat je az na konec.

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieServiceSpecification.groovy:
```diff
@@ -0,0 +1,121 @@
+                "Setup failed"
+        )
+    }
+//
```
🐛 Tady si to moc komplikujete. Vyuzijte spring, ktery vam vytvori z konfigurace datasource.  Rozbehnite pametovu DB, nebo lepe testcontainer s postgresql db. Tam se vam vytvori DB schema definovana v liquibase.

jankoscak replied:
Neudelal jste co jsem po Vas chtel. Spring boot autoconfiguruje (sam vytvori instanci DataSource), kterou muzete vyuzit v testech. Navic ten DataSource nepotrebujete. Pomocou konfigurace jste schopen pro testy vytvorit jinou schemu DB. (Mysleno tak, ze pro testy muzete mit jinej db.changelog-master.yml ve kterem budu sql navic, ktere chcete vykonat jenom pro testy)

Kleamens replied:
Takže musim vytvorit testcontainer ktery bude ke sve inicializaci pouzivat jenom liquibase a bude navic mit sql na populace database pro každy test?

jankoscak replied:
Ano. K populaci DB muzete vyuzit taky liquibase. Navic neni potreba zakazdym vytvaret a mazat DB schema. Uplne staci po testovaci metoda dostat DB do puvodniho stavu.

Kleamens replied:
Upravil jsem to, testy ted bezi v test containeru ktrey incializuje a populuje db pomoci liquibase. Po kazdem testu se db vraci do puvodniho stavu

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieServiceSpecification.groovy:
```diff
@@ -0,0 +1,121 @@
+            List<Movie> newMovieList = movieService.getAllMovies()
+            int moviesListSize = newMovieList.size()
+        then:
+            moviesListSize == originalMoviesListSize+1
```
🐛 spis nez pocet bych skontrolovatl jestli je mezi filmami ten novej vytvorenej.

jankoscak commented:
Prosim rebasnite tuhle vetev.

Kleamens commented:
Provedl jsem rebase, strašně se omlouvam nepřšlo mi ozanemeni

jankoscak commented:
> Provedl jsem rebase, strašně se omlouvam nepřšlo mi ozanemeni

Nerebasol jste to. (Nebo spatn). Mel bych videt 2 commity so zmenami. Vidim vsak vsechny commity z prvniho pull requestu (code-review) A pak tam vidim nejakej merge commit.

Kleamens commented:
Pravě udělal jsem rebase, už tam je.

jankoscak commented:
> Pravě udělal jsem rebase, už tam je.

Nevim co jste udelal, ale vysledek je, ze neni to v stave aky pozadujem. Tuhle vetev mate rebasnut vuci main/master. Tym by se meli odstranit duplicitne commity. Duplicitnymi commitami myslim ty, ktere byly v predchazejici branchi. V tehle branchi by meli byt jenom ty commity, ktere se tykaji spocku.

![image](https://github.com/user-attachments/assets/25388993-bfc4-4344-b674-e17242af8032)

Ani jeden commit so screenu nema byt v tehle branchi. 

![image](https://github.com/user-attachments/assets/fa4916e0-3c78-4889-a30f-a6b53a8b8412)

Jenom tyhlecty 2 tady maji ostat.

Kleamens committed:
fix: data creation

Kleamens commented:
Uprvail jsem to

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieServiceSpecification.groovy:
```diff
@@ -0,0 +1,121 @@
+@SpringBootTest(classes = [MoviesApiApplication])
+@ActiveProfiles("test")
+@Sql(value = "/test-data/setup.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_CLASS,config = @SqlConfig(transactionMode = SqlConfig.TransactionMode.ISOLATED))
+@Sql(value = "/test-data/final-cleanup.sql", executionPhase = Sql.ExecutionPhase.AFTER_TEST_CLASS,config = @SqlConfig(transactionMode = SqlConfig.TransactionMode.ISOLATED))
```
:bug: Pouzivate liquibase. Proc si pro testy vytvarite zvlast DB schemu?

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieServiceSpecification.groovy:
```diff
@@ -0,0 +1,121 @@
+        )
+    }
+
+    void setupSpec(){
+    }
+
+    void cleanupSpec(){
+    }
```
:bug:  Kdyz tady nic nedelate, tak proc to tady mate?

Kleamens committed:
feat: using testcontainers in tests

Removed old JUnit tests + tests now run in a testcontainer

Kleamens committed:
fix: check for specific object in tests

The tests now check for specific objects' creation/deletion

Kleamens committed:
fix: excessive schema creation

Kleamens committed:
feat: tests now use liquibase for population

Instead of executing the population within the specification , the db population and intialization is done entirely by liquibase

jankoscak commented on moviesAPI/src/test/resources/test-migrations/000-director_create.xml:
🐛 To budete pak pri kazdem pridani nove tabulky a nebo zmeny DB pridavat tuhle zmenu druhykrat do testu? Upravte to tak, ze pouzijete migracne skripty pouzite v aplikace a v testech navic k tomu pridate to co je potrebne vykonat jenom v testech.

Kleamens replied:
Mate rozbehnuty docker container?

jankoscak replied:
Ano docker container mam rozbehnuty. Nevim jak by rozbehnuti docker containeru melo souviset s tym, ze Vam to nejde zkompilovat. 

Kdyz jsi opravim chybu co tam mate a uspesne zkopmpiluji. Tak testy dopadnou takhle:
![image](https://github.com/user-attachments/assets/84871cdb-ebf4-4e1f-a800-1e8c8533abc5)

Kleamens replied:
Upravil jsem to, vsak nejsem si jisty proc ty testy neprochazeji jelikoz zkousel jsem je na nekolika pocitacich a prochazeli bez zadnych problemu.

jankoscak replied:
Podivejte sa na `Mocks.groovy`.

Kleamens replied:
![image](https://github.com/user-attachments/assets/df2aede8-afa0-41ba-8a6d-a1c32194622c)
![image](https://github.com/user-attachments/assets/16fac886-e68f-4ecf-8b74-f6b30bbfc355)

Chcete rict ze Instant.now pusobi problem?

jankoscak replied:
Ne `Instant` nepusobi problem. Screen, ze vam testy prochazi je k nicemu. Testy se muzou spoustet v ruznem poradi na ruznych strojich. Ty testy mate mezi sebou zavisle a u vas se pousti v takovem poradi, ze je vsechno ok. 

Mocks je spatne z toho pohladu, ze vytvorite jednu instanci `Director` a `Movie` (kedze to mate ako konstanty static)

Pak nepouzivate `Transactional` na mista, kde by ste mel.

jankoscak replied:
Pohli ste s tym nejako?

Kleamens replied:
Ano, omlouvam se ze nepushnul jsem update dřive.

jankoscak reviewed (changes requested):
Ty testy neprochazi. A navic vam nejde zkompilovat aplikace.

Kleamens committed:
refactor: removed excessive migrations

The migrations defined in the main application are used for the main init for tests with addition of migrations specific to tests

Commit:
fix: data downloder rename

Changed the name of the file and the class which was creating a compilation error

Commit:
fix: potential fix for failed tests

Fixed the potential reason for tests failure

Commit:
fix: test dependency fix

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/director/DirectorServiceSpecification.groovy:
```diff
@@ -0,0 +1,69 @@
+
+    def setup(){
+        mock = new Mocks();
+    }
```
:art: A co kdyby jste udelal spolecneho predka testu, ktery by vam poskytoval instanci `Movie` a `Director`?

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieUnitSpecification.groovy:
```diff
@@ -0,0 +1,45 @@
+
+    def "Creating a movie should create a movie"(){
+        given:
+            Movie movie = Mocks.movie
+            Director director = Mocks.director
```
:bug: Tady jste to zapomnel zmenit. Nemuzete dat neco na code review bez toho, aby jste si to sam spustil.

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieServiceSpecification.groovy:
```diff
@@ -0,0 +1,67 @@
+        (moviesListSize == originalMoviesListSize+1)&&(newMovieList.contains(movie))
+    }
+    def "Should give a list of movies"(){
+        given:
+        List<Movie> movies= movieService.getAllMovies();
+        when:
+        int moviesListSize = movies.size()
+        then:
+        moviesListSize == 2
+    }
```
:bug: `given` je misto, kde si pripravite testovaci data. V `when` spustite testovanou metodu. V `then` kontrolujete vysledek. Tenhle konkretni test se da zkratit na 2 radky s pouzitim jineho klicoveho slova ako jste tady pouzil (nove klicove slovo, ktere tady nemate)

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieServiceSpecification.groovy:
```diff
@@ -0,0 +1,67 @@
+            List<Movie> newMovieList = movieService.getAllMovies()
+            int moviesListSize = newMovieList.size()
+        then:
+        (moviesListSize == originalMoviesListSize+1)&&(newMovieList.contains(movie))
```
:bug: Tohle neni citelne. Kazda `condition`, kterou tady mate by mela byt na novem radku.

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieServiceSpecification.groovy:
```diff
@@ -0,0 +1,67 @@
+
+@SpringBootTest(classes = [MoviesApiApplication])
+@ActiveProfiles("test")
+@Slf4j
```
:bug: Tohlecto je tady proc?

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/movie/MovieServiceSpecification.groovy:
```diff
@@ -0,0 +1,67 @@
+        then:
+        (moviesListSize == originalMoviesListSize-1)&&(movieService.getMovieById(idToDelete).isEmpty())
+    }
+    @Transactional
```
:bug: To `Transactional` tady byt nema. Se to pak realne nevykona nad DB. 

----
Kdyz to date pryc, tak Vam test neprojde. Kedze mate u `Movie` `EqualsAndHashCode` na cely objekt, tak se bude lisit `lastModified` parameter. DB totiz ureze cast timestampy. Tady bude stacit, kdyz z equals vyloucite `lastModified`

jankoscak commented on moviesAPI/src/test/groovy/cz/mendelu/ea/xzirchuk/project/moviesAPI/director/DirectorUnitSpecification.groovy:
```diff
@@ -0,0 +1,28 @@
+
+
+    def "Creating a director should create a director"(){
+        given:
+            Director director = new Director(
+                    name: "Jeff",
+                    net_worth: 2.0,
+                    movieList: []
+            )
+        when:
+            String directorName = director.getName()
+            Float directorNetWorth = director.getNet_worth()
+            List<Movie> directorFilmography = director.getMovieList()
+        then:
+            directorName=="Jeff"
+            directorNetWorth==2.0
+            directorFilmography==[]
+    }
```
:bug: Tenhle test se da taky zkratit jak jsem psal na jinem miste.

Kleamens committed:
refactor: adjustments based on feedback
